api/envoy/service/network_ext_proc/v3/network_external_processor.proto

```diff
@@ -0,0 +1,183 @@
+syntax = "proto3";
+
+package envoy.service.network_ext_proc.v3;
+
+import "envoy/config/core/v3/base.proto";
+
+import "google/protobuf/struct.proto";
+
+import "xds/annotations/v3/status.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.service.network_ext_proc.v3";
+option java_outer_classname = "NetworkExternalProcessorProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/service/network_ext_proc/v3;network_ext_procv3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+option (xds.annotations.v3.file_status).work_in_progress = true;
+
+// [#protodoc-title: Network External Processing Service]
+
+// The Network External Processing filter allows an external service to dynamically
+// interact with and modify L4 network traffic passing through Envoy. Unlike the
+// HTTP External Processing filter, this service operates at the TCP/UDP level,
+// providing access to raw network data.
+//
+// The filter communicates with an external gRPC service that can:
+// * Inspect network traffic in both directions (client->server and server->client)
+// * Modify the payload data
+// * Control connection lifecycle (continue, close gracefully, or reset)
+//
+// Use cases include:
+// * Custom protocol inspection and modification
+// * Advanced traffic manipulation
+// * Security scanning and filtering
+// * Dynamic connection management
+//
+// The service uses a bidirectional gRPC stream, maintaining state throughout
+// the connection lifetime while allowing asynchronous processing.
+
+// NetworkExternalProcessor service defines the contract between Envoy and
+// external processing servers for L4 network traffic.
+service NetworkExternalProcessor {
+  // Process establishes a bidirectional stream between Envoy and the external
+  // processing server. Envoy sends ProcessingRequests containing network data
+  // and the server responds with ProcessingResponses containing processing
+  // decisions and potentially modified data.
+  //
+  // The server should handle processing timeout properly to avoid blocking
+  // network traffic for extended periods. Any uncaught exceptions will
+  // be treated as errors and will terminate the stream.
+  //
+  // Implementation note: The server should process requests in the order
+  // they are received to maintain proper sequencing of network traffic.
+  rpc Process(stream ProcessingRequest) returns (stream ProcessingResponse) {
+  }
+}
+
+// ProcessingRequest contains data sent from Envoy to the external processing server.
+// Each request contains either read data (from client) or write data (to client)
+// along with optional metadata.
+message ProcessingRequest {
+  oneof request {
```

Comment:
As per https://github.com/envoyproxy/envoy/blob/main/api/STYLE.md, we prefer not to use `oneof`, since that makes it difficult to add new fields later without breaking interoperability.

Same thing below.